keps/sig-storage/20190125-online-growing-persistent-volume-size.md

```diff
@@ -81,187 +81,25 @@ Enable users to increase the size of a PVC which is already in use (mounted). Th
-	}
-	if err == nil {
-		glog.V(4).Infof(volumeToMount.GenerateMsgDetailed("operationExecutor.VolumeFileSystemResize started", ""))
-	}
-}
-```
+The reconciler will periodically check whether any volume that's mounted in a pod (as reported by `ActualStateOfWorld`) requires an online resizing operation by reading its fields and triggering an online file system resize operation if: the volume's `PVC.Status.Capacity` in `ActualStateOfWorld` is less than its `PV.Spec.Capacity` in `DesiredStateOfWorld`, the volume has `volumeMode` `Filesystem`, and the volume is not `readOnly`.
 
-As can be seen from the above code, we add a `VolumeFileSystemResize` method to `OperationExecutor` and a `GenerateVolumeFSResizeFunc` method to `OperationGenerator`, which performs volume file system resizing operation.
+If the file system resizing operation succeeds, `PVC.Status.Capacity` is changed to the desired volume size in `PV.Spec.Capacity`, and the `PersistentVolumeClaimFileSystemResizePending` condition removed from `PVC.Status.Conditions`. The `PVC.Status.Capacity` in `ActualStateOfWorld` is changed too. The `PVC.Status.Capacity` in `DesiredStateOfWorld` will be changed by the populator the next time it reprocesses the pod. (If the same volume is mounted again to another pod before that, a no-op resize may be triggered).
 
-We also add a `MarkVolumeAsResized` to `ActualStateOfWorldMounterUpdater`, which will be invoked after the file system resizing operation succeeds. `MarkVolumeAsResized` clears the `fsResizeRequired` flag of the according `mountedPod` to promise that no more resize operation will be performed.
+If it fails, the reconciler retries.
 
-The main work of `GenerateVolumeFSResizeFunc` is simply invoking the `resizeFileSystem` method and `MarkVolumeAsResized`, it looks like this:
-
-```go
-	resizeSimpleError, resizeDetailedError := og.resizeFileSystem(volumeToMount, volumeToMount.DevicePath, deviceMountPath, volumePlugin.GetPluginName())
-	if resizeSimpleError != nil || resizeDetailedError != nil {
-		return resizeSimpleError, resizeDetailedError
-	}
-	markFSResizedErr := actualStateOfWorld.MarkVolumeAsResized(volumeToMount.PodName, volumeToMount.VolumeName)
-	if markFSResizedErr != nil {
-		// On failure, return error. Caller will log and retry.
-		return volumeToMount.GenerateError("VolumeFSResize.MarkVolumeAsResized failed", markFSResizedErr)
-	}
-	return nil, nil
-```
+It is important to note that:
 
-- If the file system resizing operation succeeds, `resizeFileSystem` will change `PVC.Status.Capacity` to the desired volume size in `PV.Spec.Capacity`, and remove the `PersistentVolumeClaimFileSystemResizePending` condition from `PVC.Status.Conditions`, then this resizing request is marked as finished.
+- We only perform online resizing for mounted volumes, it means that these volumes must exist in `ActualStateOfWorld`. This is why only the reconciler is expected to read/write PV & PVC fields via the volume in `ActualStateOfWorld`.
```

Comment:
How do we synchronize with Unmount?

Reply:
hm, now that https://github.com/kubernetes/kubernetes/pull/75213/ has merged we don't. previously operation executor's goroutinemap  prevented an unmount operation from starting while an expand operation was. I don't think there is an issue with a user unmounting a volume in the middle of an online resize, I would expect either it just silently becomes offline resize or unmount gets blocked i.e. I would expect kernel to be able deal with it. Might be difficult to test, but I'll add a note that it needs to be accounted for it.

Reply:
we still prevent concurrent operation on same volume - #75213 does nothing to change that.

Reply:
ack, i see key != name

Reply:
updated